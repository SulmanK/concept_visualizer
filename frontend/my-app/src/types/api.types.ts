/**
 * API-related TypeScript interfaces for the Concept Visualizer application.
 */

/**
 * API Error response model
 */
export interface ApiError {
  status: number;
  message: string;
  details?: string;
}

/**
 * Common API response wrapper
 */
export interface ApiResponse<T> {
  data?: T;
  error?: ApiError;
  loading: boolean;
}

/**
 * Task status type
 */
export type TaskStatus = "pending" | "processing" | "completed" | "failed";

/**
 * Task response model
 */
export interface TaskResponse {
  id: string;
  task_id?: string;
  status: TaskStatus;
  type: "generate_concept" | "refine_concept";
  result_id?: string;
  error_message?: string;
  created_at: string;
  updated_at: string;
}

/**
 * Request model for concept generation
 */
export interface PromptRequest {
  logo_description: string;
  theme_description: string;
}

/**
 * Request model for concept refinement
 */
export interface RefinementRequest {
  original_image_url: string;
  logo_description?: string;
  theme_description?: string;
  refinement_prompt: string;
  preserve_aspects: string[];
}
