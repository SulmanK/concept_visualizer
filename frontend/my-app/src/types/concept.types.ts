/**
 * Concept-related TypeScript interfaces for the Concept Visualizer application.
 */

/**
 * Color palette model containing hex color codes
 */
export interface ColorPalette {
  primary: string;
  secondary: string;
  accent: string;
  background: string;
  text: string;
  additionalColors: string[];
}

/**
 * Color variation for a concept
 */
export interface ColorVariation {
  name: string;
  colors: string[];
  image_url: string;
  image_path?: string;
  description?: string;
}

/**
 * Response model for concept generation and refinement
 */
export interface GenerationResponse {
  // Backend API schema uses snake_case, while frontend uses camelCase
  // We need to support both naming conventions for compatibility
  
  // Primary fields that come from the backend
  prompt_id: string;
  logo_description: string;
  theme_description: string;
  created_at: string;
  image_url: string;
  color_palette: ColorPalette | null;
  generation_id?: string;
  variations?: ColorVariation[];
  original_image_url?: string | null;
  refinement_prompt?: string | null;
  id?: string; // ID of the saved concept in the database
  
  // These are aliases for frontend compatibility
  // They might not be in the API response
  imageUrl?: string;
  colorPalette?: ColorPalette | null;
  generationId?: string;
  createdAt?: string;
  originalImageUrl?: string | null;
}

/**
 * Stored concept data structure for recent concepts
 */
export interface StoredConcept {
  id: string;
  title?: string;
  logo_description: string;
  theme_description: string;
  image_url: string;
  image_path?: string;
  created_at: string;
  color_variations?: ColorVariation[];
} 