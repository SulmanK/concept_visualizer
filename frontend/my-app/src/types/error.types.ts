/**
 * Error-related TypeScript interfaces for the Concept Visualizer application.
 */

/**
 * Error categories enumeration
 */
export type ErrorCategory =
  | "network"
  | "auth"
  | "validation"
  | "rateLimit"
  | "api"
  | "server"
  | "unknown";

/**
 * Basic application debug info
 */
export interface DebugInfo {
  environment: string;
  apiBaseUrl: string;
  supabaseUrl: string;
  [key: string]: string | number | boolean;
}

/**
 * Validation error structure
 */
export interface ValidationError {
  field: string;
  message: string;
}

/**
 * Rate limit data structure
 */
export interface RateLimitData {
  limit: number;
  current: number;
  period: string;
  resetAfterSeconds: number;
}

/**
 * Common application error interface
 */
export interface ApplicationError {
  message: string;
  details?: string;
  category: ErrorCategory;
  status?: number;
  validationErrors?: ValidationError[];
  limit?: number;
  current?: number;
  period?: string;
  resetAfterSeconds?: number;
}
