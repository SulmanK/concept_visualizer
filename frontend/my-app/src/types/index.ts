/**
 * Re-export all types from domain-specific files
 */

// API Types
export * from "./api.types";

// Concept Types
export * from "./concept.types";

// UI Types
export * from "./ui.types";

// Form Types
export * from "./form.types";

/**
 * Form status type
 */
export type FormStatus =
  | "idle"
  | "submitting"
  | "initiating"
  | "pending"
  | "processing"
  | "success"
  | "error";
