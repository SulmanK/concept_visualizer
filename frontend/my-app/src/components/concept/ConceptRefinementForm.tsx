import React, { useState, FormEvent, useEffect } from 'react';
import { Button } from '../ui/Button';
import { TextArea } from '../ui/TextArea';
import { Card } from '../ui/Card';
import { Input } from '../ui/Input';
import { FormStatus } from '../../types';
import { Spinner } from '../ui';
import { useTaskContext } from '../../contexts/TaskContext';

export interface ConceptRefinementFormProps {
  /**
   * Original image URL
   */
  originalImageUrl: string;
  
  /**
   * Handle form submission
   */
  onSubmit: (refinementPrompt: string, logoDescription: string, themeDescription: string, preserveAspects: string[]) => void;
  
  /**
   * Form submission status
   */
  status: FormStatus;
  
  /**
   * Error message from submission
   */
  error?: string | null;
  
  /**
   * Cancel refinement
   */
  onCancel?: () => void;
  
  /**
   * Initial logo description
   */
  initialLogoDescription?: string;
  
  /**
   * Initial theme description
   */
  initialThemeDescription?: string;

  /**
   * Custom placeholder text for refinement prompt
   */
  refinementPlaceholder?: string;

  /**
   * Default preserve aspects to pre-select
   */
  defaultPreserveAspects?: string[];

  /**
   * Whether we're refining a color variation
   */
  isColorVariation?: boolean;

  /**
   * Color information for the current variation
   */
  colorInfo?: {
    colors: string[];
    name: string;
  };

  /**
   * Whether the refinement request is being processed
   */
  isProcessing?: boolean;

  /**
   * Message to display during processing
   */
  processingMessage?: string;
}

/**
 * Form for submitting concept refinement requests
 */
export const ConceptRefinementForm: React.FC<ConceptRefinementFormProps> = ({
  originalImageUrl,
  onSubmit,
  status,
  error,
  onCancel,
  initialLogoDescription = '',
  initialThemeDescription = '',
  refinementPlaceholder = 'Describe how you want to refine this concept...',
  defaultPreserveAspects = [],
  isColorVariation = false,
  colorInfo,
  isProcessing = false,
  processingMessage = 'Processing your refinement request...',
}) => {
  const [refinementPrompt, setRefinementPrompt] = useState('');
  const [logoDescription, setLogoDescription] = useState(initialLogoDescription);
  const [themeDescription, setThemeDescription] = useState(initialThemeDescription);
  const [preserveAspects, setPreserveAspects] = useState<string[]>(defaultPreserveAspects);
  const [validationError, setValidationError] = useState<string | undefined>(undefined);
  
  // Get global task status
  const { hasActiveTask, isTaskPending, isTaskProcessing } = useTaskContext();
  
  // Update preserve aspects when defaultPreserveAspects changes
  useEffect(() => {
    setPreserveAspects(defaultPreserveAspects);
  }, [defaultPreserveAspects]);
  
  const aspectOptions = [
    { id: 'layout', label: 'Layout' },
    { id: 'colors', label: 'Colors' },
    { id: 'style', label: 'Style' },
    { id: 'symbols', label: 'Symbols/Icons' },
    { id: 'proportions', label: 'Proportions' },
    { id: 'color_scheme', label: 'Color Scheme' },
  ];
  
  const isSubmitting = status === 'submitting';
  const isSuccess = status === 'success';
  const showProcessing = isProcessing || isTaskPending || isTaskProcessing;
  
  // Check if any task is in progress
  const isTaskInProgress = hasActiveTask || isSubmitting || isSuccess || isProcessing;
  
  const validateForm = (): boolean => {
    if (!refinementPrompt.trim()) {
      setValidationError('Please provide refinement instructions');
      return false;
    }
    
    if (refinementPrompt.length < 5) {
      setValidationError('Refinement instructions must be at least 5 characters');
      return false;
    }
    
    setValidationError(undefined);
    return true;
  };
  
  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    
    if (validateForm()) {
      onSubmit(refinementPrompt, logoDescription, themeDescription, preserveAspects);
    }
  };
  
  const toggleAspect = (aspectId: string) => {
    setPreserveAspects(prev => 
      prev.includes(aspectId)
        ? prev.filter(id => id !== aspectId)
        : [...prev, aspectId]
    );
  };
  
  return (
    <Card 
      variant="gradient"
      className="max-w-xl mx-auto"
      header={
        <h2 className="text-xl font-semibold text-indigo-900">
          Refine Concept
          {isColorVariation && colorInfo && (
            <span className="ml-2 text-sm font-normal text-indigo-600">
              ({colorInfo.name})
            </span>
          )}
        </h2>
      }
    >
      {showProcessing ? (
        <div className="py-8 flex flex-col items-center text-center">
          <Spinner size="lg" className="mb-4" />
          <p className="text-indigo-700 font-medium">{processingMessage}</p>
          <p className="text-sm text-indigo-500 mt-2">
            This might take a minute. Please wait while we process your request.
          </p>
        </div>
      ) : (
        <form onSubmit={handleSubmit} className="space-y-4">
          {/* Original image thumbnail */}
          <div className="flex justify-center mb-4">
            <div className="w-40 h-40 border border-indigo-200 rounded-lg overflow-hidden shadow-sm">
              <img 
                src={originalImageUrl} 
                alt="Original concept" 
                className="w-full h-full object-cover"
              />
            </div>
          </div>
          
          {/* Color palette for variations */}
          {isColorVariation && colorInfo && colorInfo.colors.length > 0 && (
            <div className="mb-4">
              <p className="text-sm font-medium text-indigo-700 mb-2">Color Palette</p>
              <div className="flex flex-wrap gap-2">
                {colorInfo.colors.map((color, index) => (
                  <div 
                    key={index}
                    className="w-8 h-8 rounded-full border border-gray-200"
                    style={{ backgroundColor: color }}
                    title={color}
                  />
                ))}
              </div>
            </div>
          )}
          
          {/* Refinement instructions */}
          <div>
            <TextArea
              label="Refinement Instructions"
              placeholder={refinementPlaceholder}
              value={refinementPrompt}
              onChange={(e) => setRefinementPrompt(e.target.value)}
              error={validationError}
              fullWidth
              disabled={isTaskInProgress}
              helperText={isColorVariation 
                ? "Describe what you want to change while keeping the color scheme" 
                : "Be specific about what you want to change"
              }
              rows={3}
            />
          </div>
          
          {/* Optional: Updated logo description */}
          <div>
            <TextArea
              label="Updated Logo Description (Optional)"
              placeholder="Update the logo description or leave as is..."
              value={logoDescription}
              onChange={(e) => setLogoDescription(e.target.value)}
              fullWidth
              disabled={isTaskInProgress}
              helperText="Leave empty to keep original description"
              rows={2}
            />
          </div>
          
          {/* Optional: Updated theme description */}
          <div>
            <TextArea
              label="Updated Theme Description (Optional)"
              placeholder="Update the theme description or leave as is..."
              value={themeDescription}
              onChange={(e) => setThemeDescription(e.target.value)}
              fullWidth
              disabled={isTaskInProgress}
              helperText="Leave empty to keep original description"
              rows={2}
            />
          </div>
          
          {/* Preserve aspects checkboxes */}
          <div>
            <p className="text-sm font-medium text-gray-700 mb-2">Preserve from Original</p>
            <div className="flex flex-wrap gap-3">
              {aspectOptions.map((option) => (
                <label 
                  key={option.id}
                  className={`
                    inline-flex items-center px-3 py-1.5 rounded-full text-sm
                    ${preserveAspects.includes(option.id)
                      ? 'bg-indigo-100 text-indigo-800 border-indigo-300'
                      : 'bg-gray-50 text-gray-700 border-gray-200'
                    }
                    border cursor-pointer transition-colors
                    ${isTaskInProgress ? 'opacity-50 cursor-not-allowed' : 'hover:bg-indigo-50'}
                  `}
                >
                  <input
                    type="checkbox"
                    className="sr-only"
                    checked={preserveAspects.includes(option.id)}
                    onChange={() => !isTaskInProgress && toggleAspect(option.id)}
                    disabled={isTaskInProgress}
                  />
                  {option.label}
                </label>
              ))}
            </div>
            <p className="text-xs text-gray-500 mt-1">
              Select elements you want to keep from the original concept
            </p>
          </div>
          
          {/* Form buttons */}
          <div className="flex justify-between pt-2">
            {onCancel && (
              <Button 
                type="button" 
                variant="secondary" 
                onClick={onCancel}
                disabled={isTaskInProgress}
              >
                Cancel
              </Button>
            )}
            
            <Button 
              type="submit" 
              variant="primary"
              disabled={isTaskInProgress}
            >
              {isSubmitting ? 'Processing...' : 
               hasActiveTask ? 'Task already in progress...' : 
               'Refine Concept'}
            </Button>
          </div>
        </form>
      )}
    </Card>
  );
};

export default ConceptRefinementForm; 