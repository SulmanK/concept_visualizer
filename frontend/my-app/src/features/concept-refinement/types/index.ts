/**
 * Types specific to the Concept Refinement feature
 */
import { GenerationResponse } from '../../../types/concept.types';
import { FormStatus } from '../../../types/ui.types';

/**
 * Original concept model
 */
export interface OriginalConcept {
  imageUrl: string;
  logoDescription?: string;
  themeDescription?: string;
}

/**
 * Props for the RefinementForm component
 */
export interface RefinementFormProps {
  originalImageUrl: string;
  onSubmit: (
    refinementPrompt: string,
    logoDescription: string,
    themeDescription: string,
    preserveAspects: string[]
  ) => void;
  status: FormStatus;
  error?: string;
  onCancel: () => void;
  initialLogoDescription?: string;
  initialThemeDescription?: string;
}

/**
 * Props for the RefinementActions component
 */
export interface RefinementActionsProps {
  onReset: () => void;
  onCreateNew: () => void;
}

/**
 * Props for the ComparisonView component
 */
export interface ComparisonViewProps {
  originalConcept: OriginalConcept;
  refinedConcept: GenerationResponse;
  onColorSelect?: (color: string) => void;
  onRefineRequest?: () => void;
} 