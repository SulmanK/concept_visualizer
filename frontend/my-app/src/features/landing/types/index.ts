/**
 * Types specific to the Landing page feature
 */
import { GenerationResponse } from '../../../types/concept.types';
import { FormStatus } from '../../../types/ui.types';

/**
 * Step model for the How It Works section
 */
export interface HowItWorksStep {
  number: number;
  title: string;
  description: string;
}

/**
 * Props for the HowItWorks component
 */
export interface HowItWorksProps {
  steps: HowItWorksStep[];
  onGetStarted: () => void;
}

/**
 * Recent concept model for display in RecentConcepts section
 */
export interface RecentConcept {
  id: string;
  title: string;
  description: string;
  colors: string[];
  gradient: {
    from: string;
    to: string;
  };
  initials: string;
}

/**
 * Props for the ResultsSection component
 */
export interface ResultsSectionProps {
  result: GenerationResponse;
  onReset: () => void;
  selectedColor: string | null;
  onColorSelect: (color: string) => void;
}

/**
 * Props for the ConceptFormSection component
 */
export interface ConceptFormSectionProps {
  onSubmit: (logoDescription: string, themeDescription: string) => void;
  status: FormStatus;
  error?: string;
  onReset: () => void;
}

/**
 * Props for the RecentConceptsSection component
 */
export interface RecentConceptsSectionProps {
  concepts: RecentConcept[];
  onEdit: (id: string) => void;
  onViewDetails: (id: string) => void;
} 