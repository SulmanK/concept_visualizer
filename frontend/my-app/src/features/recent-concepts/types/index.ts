/**
 * Types specific to the Recent Concepts feature
 */
import { StoredConcept, ColorVariation } from '../../../types/concept.types';

/**
 * Props for ConceptCard component
 */
export interface ConceptCardProps {
  concept: StoredConcept;
}

/**
 * Context state for recent concepts
 */
export interface RecentConceptsState {
  recentConcepts: StoredConcept[];
  loadingConcepts: boolean;
  errorLoadingConcepts: string | null;
  refreshConcepts: () => void;
} 