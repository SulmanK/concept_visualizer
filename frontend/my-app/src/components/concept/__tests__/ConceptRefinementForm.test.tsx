import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { ConceptRefinementForm } from '../ConceptRefinementForm';
import { FormStatus } from '../../../types';

describe('ConceptRefinementForm Component', () => {
  // Mock handlers and props
  const mockSubmit = jest.fn();
  const mockCancel = jest.fn();
  const originalImageUrl = 'https://example.com/original-concept.png';
  const initialLogoDescription = 'Initial logo description';
  const initialThemeDescription = 'Initial theme description';
  
  // Reset mocks before each test
  beforeEach(() => {
    jest.clearAllMocks();
  });
  
  // Basic rendering test
  test('renders form fields and buttons', () => {
    render(
      <ConceptRefinementForm 
        originalImageUrl={originalImageUrl}
        onSubmit={mockSubmit} 
        status="idle" 
        onCancel={mockCancel}
        initialLogoDescription={initialLogoDescription}
        initialThemeDescription={initialThemeDescription}
      />
    );
    
    // Check for card title - be more specific to get the heading, not the button
    const titleHeading = screen.getByRole('heading', { name: 'Refine Concept' });
    expect(titleHeading).toBeInTheDocument();
    
    // Check for form inputs
    const refinementLabel = screen.getByText('Refinement Instructions');
    const logoLabel = screen.getByText('Updated Logo Description (Optional)');
    const themeLabel = screen.getByText('Updated Theme Description (Optional)');
    expect(refinementLabel).toBeInTheDocument();
    expect(logoLabel).toBeInTheDocument();
    expect(themeLabel).toBeInTheDocument();
    
    // Check for the original image
    const originalImage = screen.getByAltText('Original concept');
    expect(originalImage).toBeInTheDocument();
    expect(originalImage).toHaveAttribute('src', originalImageUrl);
    
    // Check for preserve aspects section
    const preserveAspectsText = screen.getByText('Preserve Aspects (Optional)');
    expect(preserveAspectsText).toBeInTheDocument();
    
    // Check for aspect options
    const layoutOption = screen.getByText('Layout');
    const colorsOption = screen.getByText('Colors');
    const styleOption = screen.getByText('Style');
    const symbolsOption = screen.getByText('Symbols/Icons');
    const proportionsOption = screen.getByText('Proportions');
    expect(layoutOption).toBeInTheDocument();
    expect(colorsOption).toBeInTheDocument();
    expect(styleOption).toBeInTheDocument();
    expect(symbolsOption).toBeInTheDocument();
    expect(proportionsOption).toBeInTheDocument();
    
    // Check for buttons
    const cancelButton = screen.getByRole('button', { name: /Cancel/i });
    const submitButton = screen.getByRole('button', { name: /Refine Concept/i });
    expect(cancelButton).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });
  
  // Initial values test
  test('renders with initial values', () => {
    render(
      <ConceptRefinementForm 
        originalImageUrl={originalImageUrl}
        onSubmit={mockSubmit} 
        status="idle" 
        onCancel={mockCancel}
        initialLogoDescription={initialLogoDescription}
        initialThemeDescription={initialThemeDescription}
      />
    );
    
    // Find the textareas by their labels
    const logoLabel = screen.getByText('Updated Logo Description (Optional)');
    const themeLabel = screen.getByText('Updated Theme Description (Optional)');
    
    // Find the textareas via their parent elements
    const logoTextarea = logoLabel.closest('div')?.querySelector('textarea');
    const themeTextarea = themeLabel.closest('div')?.querySelector('textarea');
    
    // Check that they have the initial values
    expect(logoTextarea).toHaveValue(initialLogoDescription);
    expect(themeTextarea).toHaveValue(initialThemeDescription);
  });
  
  // Form validation test - empty refinement prompt
  test('validates form and prevents submission with empty refinement prompt', () => {
    render(
      <ConceptRefinementForm 
        originalImageUrl={originalImageUrl}
        onSubmit={mockSubmit} 
        status="idle" 
        onCancel={mockCancel}
      />
    );
    
    // Find and click the submit button without filling out the refinement prompt
    const submitButton = screen.getByRole('button', { name: /Refine Concept/i });
    fireEvent.click(submitButton);
    
    // Check for validation error message
    const error = screen.getByText('Please provide refinement instructions');
    expect(error).toBeInTheDocument();
    
    // Verify that the submission handler was not called
    expect(mockSubmit).not.toHaveBeenCalled();
  });
  
  // Form validation test - refinement prompt too short
  test('validates form and prevents submission with refinement prompt that is too short', () => {
    render(
      <ConceptRefinementForm 
        originalImageUrl={originalImageUrl}
        onSubmit={mockSubmit} 
        status="idle" 
        onCancel={mockCancel}
      />
    );
    
    // Find the refinement prompt textarea by its label
    const refinementLabel = screen.getByText('Refinement Instructions');
    const refinementTextarea = refinementLabel.closest('div')?.querySelector('textarea');
    
    // Fill out the refinement prompt with a value that is too short
    if (refinementTextarea) {
      fireEvent.change(refinementTextarea, { target: { value: 'Test' } });
    }
    
    // Find and click the submit button
    const submitButton = screen.getByRole('button', { name: /Refine Concept/i });
    fireEvent.click(submitButton);
    
    // Check for validation error message
    const error = screen.getByText('Refinement instructions must be at least 5 characters');
    expect(error).toBeInTheDocument();
    
    // Verify that the submission handler was not called
    expect(mockSubmit).not.toHaveBeenCalled();
  });
  
  // Successful form submission test - with no preserved aspects
  test('submits form with valid refinement prompt and no preserved aspects', () => {
    render(
      <ConceptRefinementForm 
        originalImageUrl={originalImageUrl}
        onSubmit={mockSubmit} 
        status="idle" 
        onCancel={mockCancel}
        initialLogoDescription={initialLogoDescription}
        initialThemeDescription={initialThemeDescription}
      />
    );
    
    // Find the refinement prompt textarea by its label
    const refinementLabel = screen.getByText('Refinement Instructions');
    const refinementTextarea = refinementLabel.closest('div')?.querySelector('textarea');
    
    // Fill out the refinement prompt with a valid value
    if (refinementTextarea) {
      fireEvent.change(refinementTextarea, { target: { value: 'Make the logo more minimalist and modern' } });
    }
    
    // Find and click the submit button
    const submitButton = screen.getByRole('button', { name: /Refine Concept/i });
    fireEvent.click(submitButton);
    
    // Verify that the submission handler was called with the correct values
    expect(mockSubmit).toHaveBeenCalledWith(
      'Make the logo more minimalist and modern',
      initialLogoDescription,
      initialThemeDescription,
      []
    );
  });
  
  // Successful form submission test - with preserved aspects
  test('submits form with valid refinement prompt and selected preserved aspects', () => {
    render(
      <ConceptRefinementForm 
        originalImageUrl={originalImageUrl}
        onSubmit={mockSubmit} 
        status="idle" 
        onCancel={mockCancel}
        initialLogoDescription={initialLogoDescription}
        initialThemeDescription={initialThemeDescription}
      />
    );
    
    // Find the refinement prompt textarea by its label
    const refinementLabel = screen.getByText('Refinement Instructions');
    const refinementTextarea = refinementLabel.closest('div')?.querySelector('textarea');
    
    // Fill out the refinement prompt with a valid value
    if (refinementTextarea) {
      fireEvent.change(refinementTextarea, { target: { value: 'Make the logo more minimalist and modern' } });
    }
    
    // Select some preserve aspects checkboxes
    const colorsCheckbox = screen.getByText('Colors').previousElementSibling;
    const styleCheckbox = screen.getByText('Style').previousElementSibling;
    
    if (colorsCheckbox && styleCheckbox) {
      fireEvent.click(colorsCheckbox);
      fireEvent.click(styleCheckbox);
    }
    
    // Find and click the submit button
    const submitButton = screen.getByRole('button', { name: /Refine Concept/i });
    fireEvent.click(submitButton);
    
    // Verify that the submission handler was called with the correct values
    expect(mockSubmit).toHaveBeenCalledWith(
      'Make the logo more minimalist and modern',
      initialLogoDescription,
      initialThemeDescription,
      ['colors', 'style']
    );
  });
  
  // Test aspect selection toggle
  test('toggles preserve aspect when checkbox is clicked', () => {
    render(
      <ConceptRefinementForm 
        originalImageUrl={originalImageUrl}
        onSubmit={mockSubmit} 
        status="idle" 
        onCancel={mockCancel}
      />
    );
    
    // Find the colors checkbox
    const colorsCheckbox = screen.getByText('Colors').previousElementSibling;
    
    // Initially, it should not be checked
    expect(colorsCheckbox).not.toBeChecked();
    
    // Click to select it
    if (colorsCheckbox) {
      fireEvent.click(colorsCheckbox);
    }
    
    // Now it should be checked
    expect(colorsCheckbox).toBeChecked();
    
    // Click to unselect it
    if (colorsCheckbox) {
      fireEvent.click(colorsCheckbox);
    }
    
    // Now it should not be checked again
    expect(colorsCheckbox).not.toBeChecked();
  });
  
  // Loading state test
  test('displays loading state when status is submitting', () => {
    render(
      <ConceptRefinementForm 
        originalImageUrl={originalImageUrl}
        onSubmit={mockSubmit} 
        status="submitting" 
        onCancel={mockCancel}
      />
    );
    
    // Check for the loading button text
    const loadingButton = screen.getByText('Refining...');
    expect(loadingButton).toBeInTheDocument();
    
    // Verify that the submit button is disabled
    const submitButton = loadingButton.closest('button');
    expect(submitButton).toBeDisabled();
    
    // Verify that the textareas are disabled
    const textareas = screen.getAllByRole('textbox');
    textareas.forEach(textarea => {
      expect(textarea).toBeDisabled();
    });
    
    // Verify that the checkboxes are disabled
    const checkboxes = screen.getAllByRole('checkbox');
    checkboxes.forEach(checkbox => {
      expect(checkbox).toBeDisabled();
    });
  });
  
  // Error state test
  test('displays error message when provided', () => {
    const errorMessage = 'Failed to refine concept';
    
    render(
      <ConceptRefinementForm 
        originalImageUrl={originalImageUrl}
        onSubmit={mockSubmit} 
        status="error" 
        error={errorMessage}
        onCancel={mockCancel}
      />
    );
    
    // Check for the error message
    const errorElement = screen.getByText(errorMessage);
    expect(errorElement).toBeInTheDocument();
  });
  
  // Cancel button test
  test('calls onCancel when cancel button is clicked', () => {
    render(
      <ConceptRefinementForm 
        originalImageUrl={originalImageUrl}
        onSubmit={mockSubmit} 
        status="idle" 
        onCancel={mockCancel}
      />
    );
    
    // Find and click the cancel button
    const cancelButton = screen.getByRole('button', { name: /Cancel/i });
    fireEvent.click(cancelButton);
    
    // Verify that the cancel handler was called
    expect(mockCancel).toHaveBeenCalled();
  });
  
  // Success state test - form inputs should be disabled
  test('disables form inputs when status is success', () => {
    render(
      <ConceptRefinementForm 
        originalImageUrl={originalImageUrl}
        onSubmit={mockSubmit} 
        status="success" 
        onCancel={mockCancel}
      />
    );
    
    // Check that the form inputs are disabled
    const textareas = screen.getAllByRole('textbox');
    textareas.forEach(textarea => {
      expect(textarea).toBeDisabled();
    });
    
    // Check that the checkboxes are disabled
    const checkboxes = screen.getAllByRole('checkbox');
    checkboxes.forEach(checkbox => {
      expect(checkbox).toBeDisabled();
    });
    
    // Check that the submit button is disabled
    const submitButton = screen.getByRole('button', { name: /Refine Concept/i });
    expect(submitButton).toBeDisabled();
  });
  
  // Snapshot test
  test('matches snapshot', () => {
    const { container } = render(
      <ConceptRefinementForm 
        originalImageUrl={originalImageUrl}
        onSubmit={mockSubmit} 
        status="idle" 
        onCancel={mockCancel}
        initialLogoDescription={initialLogoDescription}
        initialThemeDescription={initialThemeDescription}
      />
    );
    
    expect(container.firstChild).toMatchSnapshot();
  });
}); 