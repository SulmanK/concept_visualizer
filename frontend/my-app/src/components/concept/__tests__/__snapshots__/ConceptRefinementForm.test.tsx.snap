// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConceptRefinementForm Component > matches snapshot 1`] = `
<div
  class="card-gradient max-w-xl mx-auto"
>
  <div
    class="px-4 py-3 sm:px-6 border-b border-indigo-200 bg-indigo-50/50"
  >
    <h2
      class="text-xl font-semibold text-dark-900"
    >
      Refine Concept
    </h2>
  </div>
  <div
    class="p-4 sm:p-6"
  >
    <form
      class="space-y-4"
    >
      <div
        class="flex justify-center mb-4"
      >
        <div
          class="w-40 h-40 border border-dark-200 rounded-lg overflow-hidden shadow-sm"
        >
          <img
            alt="Original concept"
            class="w-full h-full object-cover"
            src="https://example.com/original-concept.png"
          />
        </div>
      </div>
      <div>
        <div
          class="w-full"
        >
          <label
            class="label"
            for="textarea-2j3cef1at"
          >
            Refinement Instructions
          </label>
          <textarea
            aria-describedby="textarea-2j3cef1at-helper"
            aria-invalid="false"
            class="input resize-y  w-full"
            id="textarea-2j3cef1at"
            placeholder="Describe how you want to refine this concept..."
            rows="3"
          />
          <p
            class="helper-text"
            id="textarea-2j3cef1at-helper"
          >
            Be specific about what you want to change
          </p>
        </div>
      </div>
      <div>
        <div
          class="w-full"
        >
          <label
            class="label"
            for="textarea-wajfge0re"
          >
            Updated Logo Description (Optional)
          </label>
          <textarea
            aria-describedby="textarea-wajfge0re-helper"
            aria-invalid="false"
            class="input resize-y  w-full"
            id="textarea-wajfge0re"
            placeholder="Update the logo description or leave as is..."
            rows="2"
          >
            Initial logo description
          </textarea>
          <p
            class="helper-text"
            id="textarea-wajfge0re-helper"
          >
            Leave empty to keep original description
          </p>
        </div>
      </div>
      <div>
        <div
          class="w-full"
        >
          <label
            class="label"
            for="textarea-18wamhien"
          >
            Updated Theme Description (Optional)
          </label>
          <textarea
            aria-describedby="textarea-18wamhien-helper"
            aria-invalid="false"
            class="input resize-y  w-full"
            id="textarea-18wamhien"
            placeholder="Update the theme description or leave as is..."
            rows="2"
          >
            Initial theme description
          </textarea>
          <p
            class="helper-text"
            id="textarea-18wamhien-helper"
          >
            Leave empty to keep original description
          </p>
        </div>
      </div>
      <div>
        <p
          class="label mb-2"
        >
          Preserve Aspects (Optional)
        </p>
        <div
          class="flex flex-wrap gap-3"
        >
          <label
            class="flex items-center space-x-2 cursor-pointer"
          >
            <input
              class="rounded text-primary-600 focus:ring-primary-500 h-4 w-4"
              type="checkbox"
            />
            <span
              class="text-sm text-dark-700"
            >
              Layout
            </span>
          </label>
          <label
            class="flex items-center space-x-2 cursor-pointer"
          >
            <input
              class="rounded text-primary-600 focus:ring-primary-500 h-4 w-4"
              type="checkbox"
            />
            <span
              class="text-sm text-dark-700"
            >
              Colors
            </span>
          </label>
          <label
            class="flex items-center space-x-2 cursor-pointer"
          >
            <input
              class="rounded text-primary-600 focus:ring-primary-500 h-4 w-4"
              type="checkbox"
            />
            <span
              class="text-sm text-dark-700"
            >
              Style
            </span>
          </label>
          <label
            class="flex items-center space-x-2 cursor-pointer"
          >
            <input
              class="rounded text-primary-600 focus:ring-primary-500 h-4 w-4"
              type="checkbox"
            />
            <span
              class="text-sm text-dark-700"
            >
              Symbols/Icons
            </span>
          </label>
          <label
            class="flex items-center space-x-2 cursor-pointer"
          >
            <input
              class="rounded text-primary-600 focus:ring-primary-500 h-4 w-4"
              type="checkbox"
            />
            <span
              class="text-sm text-dark-700"
            >
              Proportions
            </span>
          </label>
        </div>
      </div>
      <div
        class="flex justify-end space-x-3 pt-2"
      >
        <button
          class="inline-flex items-center justify-center font-medium transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary/50 disabled:opacity-60 disabled:cursor-not-allowed border border-indigo-300 text-indigo-700 bg-white hover:bg-indigo-50 hover:text-primary-dark text-sm px-4 py-2 rounded-lg "
          type="button"
        >
          Cancel
        </button>
        <button
          class="inline-flex items-center justify-center font-medium transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary/50 disabled:opacity-60 disabled:cursor-not-allowed bg-gradient-primary text-white shadow-modern hover:shadow-modern-hover hover:brightness-105 text-sm px-4 py-2 rounded-lg "
          type="submit"
        >
          Refine Concept
        </button>
      </div>
    </form>
  </div>
</div>
`;
